File: apps/felix_web/assets/js/phasertron/utils/ai/blackboard.ts
export interface IBlackboardObject {}

export interface IBlackboard {
  set: (key: string, value: any) => void;
  get: <T>(key: string, fallback?: T) => T | null;
  tagObject: (tags: string[], gameObject: IBlackboardObject) => void;
  removeObjectTags: (tags: string[], gameObject: IBlackboardObject) => void;
  getTagged: (tag: string) => IBlackboardObject[];
}

export default class Blackboard implements IBlackboard {
  private data: { [key: string]: any } = {};
  public set = (key: string, value: any) => {
    this.data[key] = value;
  };
  public get = <T>(key: string, fallback?: T): T | null => {
    return (this.data[key] as T) ?? fallback ?? null;
  };

  private taggedObjects: { [tag: string]: IBlackboardObject[] } = {};

  public tagObject = (tags: string[], gameObject: IBlackboardObject) => {
    let tag;
    for (let i = 0; i < tags.length; i++) {
      tag = tags[i];
      this.taggedObjects[tag] = this.taggedObjects[tag] || [];
      this.taggedObjects[tag].push(gameObject);
    }
  };

  public removeObjectTags = (tags: string[], gameObject: IBlackboardObject) => {
    let tag;
    for (let i = 0; i < tags.length; i++) {
      tag = tags[i];
      if (!this.taggedObjects[tag]) {
        continue;
      }
      this.taggedObjects[tag] = this.taggedObjects[tag].filter(
        (x) => x !== gameObject
      );
    }
  };

  public getTagged = (tag: string): IBlackboardObject[] => {
    return this.taggedObjects[tag] || [];
  };
}

File: apps/felix_web/assets/js/phasertron/utils/lua/Blackboard.lua
Blackboard = {}
Blackboard.__index = Blackboard

function Blackboard:new()
    local newBlackboard = {}
    setmetatable(newBlackboard, Blackboard)
    newBlackboard.data = {}
    newBlackboard.taggedObjects = {}
    return newBlackboard
end

function Blackboard:set(key, value)
    print("\t blackboard setting " .. key)
    if (value == nil) then
        throw("setting " .. key .. " as nil")
    end
    self.data[key] = value
end

function Blackboard:get(key, fallback)
    return self.data[key] or fallback or nil
end

function Blackboard:tagObject(tags, gameObject)
    -- tags might just be a string - convert to table if so
    if type(tags) == "string" then
        tags = {tags}
    end

    for i, tag in ipairs(tags) do
        self.taggedObjects[tag] = self.taggedObjects[tag] or {}
        table.insert(self.taggedObjects[tag], gameObject)

        print("Added object to " .. tag .. " category (" .. #self.taggedObjects[tag] .. ")")
    end
end

function Blackboard:removeObjectTags(tags, gameObject)
    for i, tag in ipairs(tags) do
        if  self.taggedObjects[tag] then
            for j, object in ipairs(self.taggedObjects[tag]) do
                if object == gameObject then
                    table.remove(self.taggedObjects[tag], j)
                end
            end
        end
    end
end

function Blackboard:getTagged(tag)
    return self.taggedObjects[tag] or nil
end

return Blackboard
